### ISSUE TITLE
`redirect_to(request.referrer || root_path)` on Rails 7

### ISSUE THREAD

**Issue by joemasilotti** [varvet/pundit#727]
In the README and docs the following method is recommended to recover from unauthorized access.

```
def user_not_authorized
  flash[:alert] = I18n.t("pundit.errors.unauthorized")
  redirect_to(request.referrer || root_path)
end
```

I just upgraded to Rails 7 and this tried to redirect the user to `google.com`. I only know because [I received an error](https://app.honeybadger.io/fault/93997/9b66f4d7195c723571e05fb47bd17d67) that I should be using `allow_other_host: true` to the redirection!

---

So, my question is if `request.referrer` is the best approach here. Should instead it be recommended to `redirect_back(fallback_location: root_path)? Or something else? Or maybe I'm missing something entirely!

**Comment by dgmstuart** [comment]
Hi - thanks for the question.

I'm not sure I understand your use-case: why would you redirect the user to an external site (like google.com) when authorization fails? This will mean that they never see the flash message. 

The expected use case is that you redirect the user back to the page from which they tried to log in, and show the flash message there. If the documentation is unclear, maybe we can make it clearer?

`redirect_to request.referer` is [just a wrapper around `redirect_to request.referer`](https://github.com/rails/rails/blob/de53ba56cab69fb9707785a397a59ac4aaee9d6f/actionpack/lib/action_controller/metal/redirecting.rb#L95-L128), so you'd still have to pass `allow_other_host: true`

In conclusion: yes we should probably be suggesting `redirect_back(fallback_location: root_path)` in the README, but if you want to redirect to an external site, you'll need to explicitly allow that.

**Comment by dgmstuart** [comment]
@joemasilotti still not sure we're on the same page:

These two are entirely identical in terms of what they do, right?

```ruby
redirect_to(request.referrer || root_path)
```

```ruby
redirect_back(fallback_location: root_path)
```

If you look at the source code, it's just syntactic sugar: the implementation of the second line is basically the first line?

I do think the second line is clearer and what we should say in the README - I just want to be clear between us that it's _exactly_ the same behaviour.

**Comment by joemasilotti** [comment]
Ah, got it. I thought `redirect_back` did something fancy.

So, even though this issue isn't related to pundit, do you have any advice on how to _not_ redirect if the referrer is external?

**Comment by dgmstuart** [comment]
@joemasilotti I'm not sure I understand the question - isn't the `ActionController::Redirecting::UnsafeRedirectError` (which you described in your original post) the mechanism which prevents you from redirecting to external referrers?

**Comment by joemasilotti** [comment]
It is! But it raises an exception. I'd rather not raise the error and redirect to `root_path` in that case.

Or is the recommended approach to `rescue ActionController::Redirecting::UnsafeRedirectError`?

**Comment by Burgestrand** [comment]
> It is! But it raises an exception. I'd rather not raise the error and redirect to `root_path` in that case.
> 
> Or is the recommended approach to `rescue ActionController::Redirecting::UnsafeRedirectError`?

It [looks like it from the commit in rails](https://github.com/rails/rails/commit/c3758a71af949db849d5b7f176677653e4e4fae9), and I think that's the approach I'd take too.

**Comment by joemasilotti** [comment]
Ah, perfect! Thanks so much for the help with this.
